### PR TITLE
[utils] adding 'LookupTable' class and use it in CLI

### PR DESCRIPTION
**Pull request by abtink**
This commit adds `LookupTable` module which provides helper methods to
perform binary search in a generic sorted lookup table for specific
entry matching a given name string. It also provides `constexpr`
methods to check (at build-time) whether a lookup table array is
sorted. This commit also updates CLI modules to use the `LookupTable`
methods for command lookup. It also adds a unit test for the lookup
table.

### CODE REVIEW

File: src/core/utils/lookup_table.hpp [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+    template <class EntryType> static constexpr bool IsSorted(const EntryType *aTable, uint16_t aLength)
+    {
+        return (aLength <= 1) ? true
+                              : AreInOrder(aTable[0].mName, aTable[1].mName) && IsSorted(aTable + 1, aLength - 1);
+    }
+
+    constexpr static bool AreInOrder(const char *aFirst, const char *aSecond)
+    {
+        return (*aFirst < *aSecond) ? true : ((*aFirst > *aSecond) ? false : AreInOrder(aFirst + 1, aSecond + 1));
```
Comment (jwhui) [inline src/core/utils/lookup_table.hpp:136]: Do we properly handle strings that are equal? In other words, what if two adjacent strings are the same?

Comment (abtink) [reply src/core/utils/lookup_table.hpp:136]: Good point. I will update it (will add a case for it in unit test as well).

Comment (abtink) [reply src/core/utils/lookup_table.hpp]: Added this on new push (we ensure that there are no duplicate names in the table).

**Review by jwhui — APPROVED** [review]

**Comment by eujean** [comment]
Getting a few undefined references when building an example for cc1352:

> src/core/utils/lookup_table.hpp:117: undefined reference to `ot::Utils::LookupTable::Find(char const*, void const*, unsigned short, unsigned short, char const* (&)(void const*))'

**Comment by jwhui** [comment]
> Getting a few undefined references when building an example for cc1352:
> 
> > src/core/utils/lookup_table.hpp:117: undefined reference to `ot::Utils::LookupTable::Find(char const*, void const*, unsigned short, unsigned short, char const* (&)(void const*))'

@eujean , did you try running `./bootstrap` again? Can you provide the build commands that you are using? Note that the cc1352 build check is [currently passing](https://github.com/openthread/openthread/runs/1206649393?check_suite_focus=true).

**Pull request merged**